src/cgroups/cgroup_utils.c

```diff
@@ -724,3 +724,109 @@ bool mkdir_p(const char *dir, mode_t mode)
+ * @hierarchy_fd:	File descriptor for the hierarchy.
+ * @cgroup:		A cgroup directory relative to @hierarchy_fd.
+ * @file:		The file in @cgroup from which to read a value.
+ * @value:		Return argument to store value read from @file.
+ *
+ * This function tries to read a valid value from @file in @cgroup in
+ * @hierarchy_fd. If it is a legacy cgroup hierarchy and we fail to find a
+ * valid value we terminate early and report an error.
+ * The cgroup2 hierarchy however, has different semantics. In a few controller
+ * files it will show the value "max" or simply leave it completely empty
+ * thereby indicating that no limit has been set for this particular cgroup.
+ * However, that doesn't mean that there's no limit. A cgroup further up the
+ * hierarchy could have a limit set that also applies to the cgroup we are
+ * interested in. So for the unified cgroup hierarchy we need to keep walking
+ * towards the cgroup2 root cgroup and try to parse a valid value.
+ */
+int cgroup_walkup_to_root(int cgroup2_root_fd, int hierarchy_fd,
+			  const char *cgroup, const char *file, char **value)
+{
+	__do_close int dir_fd = -EBADF;
+	__do_free char *val = NULL;
+	bool no_limit = false;
+
+	/* Look in our current cgroup for a valid value. */
+	dir_fd = openat(hierarchy_fd, cgroup, O_DIRECTORY | O_PATH | O_CLOEXEC);
+	if (dir_fd < 0)
+		return -errno;
+
+	val = readat_file(dir_fd, file);
+	if (is_empty_string(val) || strcmp(val, "max") == 0) {
+		no_limit = true;
+	} else if (!is_empty_string(val)) {
+		*value = move_ptr(val);
+		return 0;
+	}
+
+	/*
+	 * Legacy cgroup hierarchies should always show a valid value in the
+	 * file of the cgroup. So no need to do this upwards walking crap.
+	 */
+	if (cgroup2_root_fd < 0)
+		return -EINVAL;
+
+	free_disarm(val);
+	for (int i = 0;; i++) {
+		__do_close int cur_fd = -EBADF, new_dir_fd = -EBADF;
+		__do_free char *new_val = NULL;
+		int swap_fd;
+
+		new_dir_fd = move_fd(dir_fd);
+		cur_fd = openat(new_dir_fd, "../", O_DIRECTORY | O_PATH | O_CLOEXEC);
+		if (cur_fd < 0)
+			return -errno;
+
+		/*
+		 * We're at the root of the cgroup2 tree so stop walking
+		 * upwards.
+		 */
+		if (same_file(cgroup2_root_fd, cur_fd)) {
+			if (no_limit)
+				return 1;
+
+			return -EINVAL;
+		}
+
+		/* We found a valid value. Terminate walk. */
+		new_val = readat_file(new_dir_fd, file);
+		if (is_empty_string(new_val) || strcmp(new_val, "max") == 0) {
+			no_limit = true;
+		} else if (!is_empty_string(new_val)) {
+			*value = move_ptr(new_val);
+			return 0;
+		}
+
+		/*
+		 * Set an arbitraty hard-coded limit to prevent us from ending
+		 * up in an endless loop. There really shouldn't be any cgroup
+		 * tree that is 1000 levels deep. That would be insane in
+		 * principal and performance-wise.
+		 */
+		if (i == 1000)
+			return log_error_errno(-ELOOP, ELOOP, "To many nested cgroups or invalid mount tree. Terminating walk");
+
+		swap_fd = move_fd(cur_fd);
+		cur_fd = move_fd(new_dir_fd);
+		new_dir_fd = swap_fd;
```

Comment:
Pretty sure one of these should be using `dir_fd`, since both `cur_fd` and `new_dir_fd` are inside this scope and therefore get closed at this point I think.

Reply:
dir_fd is moved before the first openat()

Reply:
oh right, i see

Reply:
Check whether what I have now makes sense, please?